src/main/java/com/amazon/aws/iot/greengrass/semver/Range.java

```diff
@@ -0,0 +1,486 @@
+            rangeMap.put(comp.value, comp);
+        }
+        if (rangeMap.size() > 1) {
+            rangeMap.remove("");
+        }
+
+        return new ArrayList<>(rangeMap.values());
+    }
+
+    public boolean intersects(Range range) {
+        return this.set.stream().anyMatch((thisComparators) -> (isSatisfiable(thisComparators) && range.set.stream()
+                .anyMatch((rangeComparators) -> (isSatisfiable(rangeComparators) && thisComparators.stream().allMatch(
+                        (thisComparator) -> rangeComparators.stream().allMatch(thisComparator::intersects))))));
+    }
+
+    // if ANY of the sets match ALL of its comparators, then pass
+    public boolean test(SemVer version) {
+        if (version == null) {
+            return false;
+        }
+
+        for (List<Comparator> comparators : this.set) {
+            if (testSet(comparators, version)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    static class FreeFunctions {
+
+        static boolean isNullSet(Comparator c) {
+            return c.value.equals("<0.0.0-0");
+        }
+
+        static boolean isAny(Comparator c) {
+            return c.value.equals("");
+        }
+
+        // take a set of comparators and determine whether there
+        // exists a version which can satisfy it
+        static boolean isSatisfiable(List<Comparator> comparators) {
+            boolean result = true;
+            if (comparators.isEmpty()) {
+                return true;
+            }
+            List<Comparator> remainingComparators = new ArrayList<>(comparators);
+            Comparator testComparator = remainingComparators.remove(remainingComparators.size() - 1);
+
+            while (result && !remainingComparators.isEmpty()) {
+                Comparator finalTestComparator = testComparator;
+                result = remainingComparators.stream().allMatch(finalTestComparator::intersects);
+                testComparator = remainingComparators.remove(remainingComparators.size() - 1);
+            }
+
+            return result;
+        }
+
+        // comprised of xranges, tildes, stars, and gtlt's at this point.
+        // already replaced the hyphen ranges
+        // turn into a set of JUST comparators.
+        static String parseComparator(String comp) {
+            comp = replaceCarets(comp);
+            comp = replaceTildes(comp);
+            comp = replaceXRanges(comp);
+            comp = replaceStars(comp);
+            return comp;
+        }
+
+        static boolean isX(String id) {
+            return id == null || id.trim().isEmpty() || id.equalsIgnoreCase("x") || id.equals("*");
+        }
+
+        static String replaceXRanges(String comp) {
+            return Arrays.stream(S_PLUS.split(comp.trim(), -1))
+                    .map(FreeFunctions::replaceXRange)
+                    .collect(Collectors.joining(" "));
+        }
+
+        private static String replaceXRange(String s) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 19. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.